src/go/rdctl/cmd/snapshot.go

```diff
@@ -72,6 +72,10 @@ func wrapSnapshotOperation(cmd *cobra.Command, appPaths paths.Paths, resetOnFail
 	if err := wrappedFunction(); err != nil {
 		if resetOnFailure {
 			factoryreset.DeleteData(appPaths, true)
+		} else {
+			if err := ensureBackendStarted(); err != nil {
+				snapshotErrors = append(snapshotErrors, err)
```

Comment:
It looks straightforward, but you're losing the error from the call to `wrappedFunction()`. If we want to report all the errors (and why would we not), the code needs to be more like this:

```
	if err := wrappedFunction(); err != nil {
		if resetOnFailure {
			factoryreset.DeleteData(appPaths, true)
		} else {
			snapshotErrors = append(snapshotErrors, err)
			err = ensureBackendStarted()
		}
		return err
	}
```

Reply:
Do you really want to return `nil` if `ensureBackendStarted()` returned nil?

Or should the code be:

```golang
	if err := wrappedFunction(); err != nil {
		if resetOnFailure {
			factoryreset.DeleteData(appPaths, true)
		} else {
			err2 := ensureBackendStarted()
			if err2 != nil {
				snapshotErrors = append(snapshotErrors, err)
				err = err2
			}
		}
		return err
	}
```

This awkwardness comes from the fact that you are returning the last error separately from all the other errors instead of returning an `[]error` list.


I'm still confused why you are not restarting the app after the factory reset?

Reply:
> but you're losing the error from the call to `wrappedFunction()`

Reply:
> > but you're losing the error from the call to `wrappedFunction()`

Who is? That is the point of my version: I save the error from `wrappedFunction` in `snapshotErrors`, but only when `ensureBackendStarted()` returns non-nil. Otherwise the error from the wrapped function remains in `err`.

Reply:
> Or am I deluded???

No, I think you are right. I just assumed that the return value was the last error that occurred, while you always return the error from the wrapped function.

Which might actually make more sense; idk. As I wrote, the error handling in this function feels rather unconventional to me.

Reply:
The weird part is that there's a global `snapshotErrors` array that tracks encountered errors and reports them at the very end. If `wrapSnapshotOperation` returns `nil` but `len(snapshotErrors) > 0` we'll still finish in an error state